In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston

### Задание 1

In [9]:
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [3]:
from sklearn.model_selection import train_test_split

In [10]:
data = boston['data']
data.shape

(506, 13)

In [5]:
feature_name = boston['feature_names']
feature_name

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [9]:
target = boston['target']
target[:5]

array([24. , 21.6, 34.7, 33.4, 36.2])

In [10]:
X = pd.DataFrame(data, columns=feature_name)
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [11]:
Y = pd.DataFrame(target, columns=['price'])
Y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [12]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)

In [13]:
Y_train

price
5     28.7
116   21.2
45    19.3
16    23.1
468   19.1
..     ...
106   19.5
270   21.1
348   24.5
435   13.4
102   18.6

[354 rows x 1 columns]

In [14]:
from sklearn.linear_model import LinearRegression

In [15]:
lr = LinearRegression()

In [16]:
lr.fit(X_train, Y_train)

LinearRegression()

In [17]:
Y_prediction = lr.predict(X_test)
Y_prediction.shape

(152, 1)

In [18]:
check_test = pd.DataFrame({
    "Y_test": Y_test["price"],
    "Y_pred": Y_prediction.flatten(),
})

check_test.head(10)

Y_test     Y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [19]:
check_test['error'] = check_test['Y_pred'] - check_test['Y_test']
check_test.head()

Y_test     Y_pred     error
173    23.6  28.648960  5.048960
274    32.4  36.495014  4.095014
491    13.6  15.411193  1.811193
72     22.8  25.403213  2.603213
452    16.1  18.855280  2.755280

In [20]:
from sklearn.metrics import r2_score

In [31]:
r2_score_issure = r2_score(check_test["Y_pred"], check_test["Y_test"])
r2_score_issure

0.669370269149561

### Задание 2

In [22]:
from sklearn.ensemble import RandomForestRegressor

In [23]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

In [26]:
model.fit(X_train, Y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [27]:
y_pred = model.predict(X_test)
y_pred.shape

(152,)

In [29]:
check_test_1 = pd.DataFrame({
    "Y_test": Y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test_1.head(10)

Y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521
76     20.0  21.832284
316    17.8  19.895747
140    14.0  14.754118
471    19.6  21.240835
500    16.8  20.898658

In [30]:
check_test_1['error'] = check_test_1['y_pred'] - check_test_1['Y_test']
check_test_1

Y_test     y_pred     error
173    23.6  22.806412 -0.793588
274    32.4  31.131464 -1.268536
491    13.6  16.339125  2.739125
72     22.8  23.810726  1.010726
452    16.1  17.139521  1.039521
..      ...        ...       ...
441    17.1  13.521640 -3.578360
23     14.5  15.112621  0.612621
225    50.0  42.808700 -7.191300
433    14.3  15.586103  1.286103
447    12.6  16.072425  3.472425

[152 rows x 3 columns]

In [32]:
r2_score_issure_1 = r2_score(check_test_1['y_pred'], check_test_1['Y_test'])
r2_score_issure_1

0.8479049999699443

Так как в первом задании мы получили в r2_score 0.669370269149561, а в r2_score_issure_1 0.8479049999699443
это значит (как я понимаю), данное решение подходит больше и показывает куда больше результат

### Задание 3

In [33]:
help(RandomForestRegressor)

Help on class RandomForestRegressor in module sklearn.ensemble._forest:

class RandomForestRegressor(ForestRegressor)
 |  RandomForestRegressor(n_estimators=100, *, criterion='mse', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features='auto', max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, bootstrap=True, oob_score=False, n_jobs=None, random_state=None, verbose=0, warm_start=False, ccp_alpha=0.0, max_samples=None)
 |  
 |  A random forest regressor.
 |  
 |  A random forest is a meta estimator that fits a number of classifying
 |  decision trees on various sub-samples of the dataset and uses averaging
 |  to improve the predictive accuracy and control over-fitting.
 |  The sub-sample size is controlled with the `max_samples` parameter if
 |  `bootstrap=True` (default), otherwise the whole dataset is used to build
 |  each tree.
 |  
 |  Read more in the :ref:`User Guide <forest>`.
 |  
 |  Parameters
 |  ----------

In [ ]:
feature_importances = RandomForestRegressor(feature_importances_=)